dask/array/core.py

```diff
@@ -371,10 +371,7 @@ def apply_infer_dtype(func, args, kwargs, funcname, suggest_dtype="dtype", nout=
     : dtype or List of dtype
         One or many dtypes (depending on ``nout``)
     """
-    args = [
-        np.ones((1,) * x.ndim, dtype=x.dtype) if isinstance(x, Array) else x
-        for x in args
-    ]
+    args = [meta_nonempty(x) for x in args]
```

Comment:
Wouldn't using `meta_from_array` suffice here?

Reply:
My understanding is it needs to be a nonempty array for gufuncs to work. If I switch this to `meta_from_array` I get:

```
           ValueError: `dtype` inference failed in `apply_gufunc`.
           
           Please specify the dtype explicitly using the `output_dtypes` kwarg.
           
           Original error is below:
           ------------------------
           ValueError('cannot call `vectorize` on size 0 inputs unless `otypes` is set')
           
           Traceback:
           ---------
             File "/home/julia/dask/dask/array/core.py", line 377, in apply_infer_dtype
               o = func(*args, **kwargs)
             File "/home/julia/conda/envs/dask-dev/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2091, in __call__
               return self._vectorize_call(func=func, args=vargs)
             File "/home/julia/conda/envs/dask-dev/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2157, in _vectorize_call
               res = self._vectorize_call_with_signature(func, args)
             File "/home/julia/conda/envs/dask-dev/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2226, in _vectorize_call_with_signature
               raise ValueError('cannot call `vectorize` on size 0 inputs '

```

Reply:
In that case, couldn't you just replace `np.ones` by [`ones_like_safe`] in the old code? That way you wouldn't even need the new `meta_from_array` function, unless I'm missing something.